### DISCUSSION TITLE
How to have a useEffect that uses `router.push` and should only fire for `router.query.id` changes in a Next.js project

### DISCUSSION THREAD

**Discussion by cauebahia-vm** [vercel/next.js#31617] [Help] [answered]
I need to transition between 2 routes in a NextJs app like the one on the following example:

```
FROM: /some-route?id=ID_1
TO:   /some-route?id=ID_1&page=2
```

And I have a `useEffect` that does something like this:

- Should fire for any `id` change
- Makes use of `router.push` for some reason

```
import { useRouter } from 'next/router';

const { query, push } = useRouter();
const { id } = query; 

useEffect(() => {
  try {
    fetchDataFor(id);
  }
  catch(err) {
    push('/not-found');
  }
},[id,push]);
```

The problem I'm facing with this is related to the fact that the whole `router` object will change from the route update, and my effect will re-run because of the `push` method reference change, even though my `id` is not changing.

I'm currently having to check if the `id` has been updated or not by using some `useRef` variable that keeps track of the previous value. That's cumbersome.

I thought that maybe I should use a dynamic route, like:

```
/some-route/[id]
/some-route/[id]?page=2
```

But I guess the problem will remain, because of the `push` property that will keep on changing because of the querystring change.

Another option would be omitting the `push` method from the useEffect's dependency array. But that seems like an anti-pattern.

What would be the best approach to solve this issue?

The goal here is being able to use `router.push` (or at least be able to push another route somehow) inside the `useEffect` that should only run for `query.id` updates.

**Comment by jamesmosier** [comment, marked as answer]
> But that seems like an anti-pattern.

Why do you think that? If the logic within your `useEffect` doesn't rely on changes to the `push` function, I think it makes sense to omit it from the dependencies array. 

As you mentioned, your only other option is to use `useRef` and compare the old and new values of id, but seems like more to maintain and requires other devs to have more understanding of what is going on with this code. Whereas if you just omit the `push` function from the dependencies array, in the future when looking at this code it is obvious that the effect should only run when `id` changes.

**Reply by cauebahia-vm** [reply]
Thank you for your reply. Well, it might not be an anti-pattern, but it's usually not recommended. Is this a safe situation to do that? I'm not aware of the internal implementation of the `push` method and why it's recreated from scratch, every time the `route` changes. I guess it wouldn't do any harm in my situation, because of the fact that if the `id` has changed, I'll be getting a brand new `push` method reference anyway, because of the route change. But still, I would like to know if are there other options. 

See: https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528

**Reply by jamesmosier** [reply]
Here is the implementation of [`push`](https://github.com/vercel/next.js/blob/a9dd61e82685bf3bdef107fa1a77e1a569ca9e1e/packages/next/shared/lib/router/router.ts?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L864-L880).

This docs explains why omitting functions [might be a bad idea](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies). In this case, because `push` uses that `id` param that you pass in and not any other values from the component it is safe. If you see in that doc it mentions:

> If after that we still don’t use any values from the component scope, it’s safe to specify []:

(in this case `[]` is `[id]`).

So as you mentioned, the other option is to use a `ref` to compare the values, which would be fine if you did not want to exclude `push`.
